coredns/plugin/zone.go

```diff
@@ -0,0 +1,233 @@
+package kuadrant
+
+import (
+	"context"
+	"net"
+	"strconv"
+	"strings"
+
+	"github.com/coredns/coredns/plugin/file"
+	"github.com/coredns/coredns/plugin/metadata"
+	"github.com/coredns/coredns/plugin/pkg/dnsutil"
+	"github.com/coredns/coredns/request"
+	"github.com/miekg/dns"
+
+	"sigs.k8s.io/external-dns/endpoint"
+
+	"github.com/kuadrant/dns-operator/api/v1alpha1"
+)
+
+const continentGEOPrefix = "GEO-"
+
+type geodRR struct {
+	dns.RR
+	geo string
+}
+
+type rrData struct {
+	weight *int64
+	geo    *string
+}
+
+type Zone struct {
+	file   *file.Zone
+	rrData map[dns.RR]rrData
+}
+
+func NewZone(name string) *Zone {
+	z := &Zone{
+		file.NewZone(name, ""),
+		map[dns.RR]rrData{},
+	}
+
+	z.file.RRResolver = func(ctx context.Context, state request.Request, rrs []dns.RR) dns.RR {
+		log.Debugf("resolving %s in zone %s", rrs[0].Header().Name, name)
+		rrMeta := z.rrData[rrs[0]]
+		if rrMeta.geo != nil {
+			rrs = z.parseGeoAnswers(ctx, state, rrs)
+		} else if rrMeta.weight != nil {
+			rrs = z.parseWeightedAnswers(ctx, state, rrs)
+		} else {
+			//Take the first answer in the default case, not geo or weighted
+			rrs = []dns.RR{rrs[0]}
+		}
+		return rrs[0]
+	}
+
+	ns := &dns.NS{Hdr: dns.RR_Header{Name: dns.Fqdn(name), Rrtype: dns.TypeNS, Ttl: ttlSOA, Class: dns.ClassINET},
+		Ns: dnsutil.Join("ns1", name),
+	}
+	z.file.Insert(ns)
```

Comment:
haven't checked but assume this checks before adding to remove duplication concern?